### DISCUSSION TITLE
Introduce a system to suppress existing errors

### DISCUSSION THREAD

**Discussion by jfmengels** [biomejs/biome#339] [Ideas]
(contents mostly stolen from my own issue in ESLint's repo https://github.com/eslint/eslint/issues/16755)

### What problem do you want to solve?

I would like there to be a system to suppress existing errors from being reported, so that I can enable a rule as `"error"` and be notified when new ones show up.

I believe that the way that this is done today in Biome and many other linters is through different imperfect means.

### Using warnings

Warnings can help report errors without making required to be fixed.

A large problem with warnings is that as soon as there are more than a few warnings, you don't notice new ones showing up. A common practice I've seen quite often is to avoid warnings altogether, and to only use errors to avoid new problems creeping in. But that doesn't solve the problem of all the existing errors.

Also, users can too easily ignore the new errors, so in a way, the rule is enabled without being enforced when IMO the point of a linter is to enforce rules.

### Using disable comments

One can use disable comments to temporarily suppress errors, by adding a comment like `// biome-ignore lint: FIX THIS LATER <explanation>`

Disable comments can be used to enable a rule as an error early, by adding them everywhere where an error is currently reported (and that is actually something that can be automated by some linters).

But disable comments have the tendency to be hard to distinguish from other disable comments created for reasons such as false positives or disagreements on the rule, especially when there is no enforced need to add a message on the comment. Meaning that once you decide to tackle the existing errors, they can be hard to detect (or to distinguish from ones that are disabled for good reasons).

They also "pollute" the codebase in a way that is quite visible, and makes users numb to the fact of using disable comments.

### Ignoring parts of the project

It is also possible to simply disable the rule in each file that is currently reporting errors, either through manually configuring the rule in the configuration, or by adding a disable comment at the top of the file that disables the rule for the entire file.

This has multiple downsides:
- While new errors are enforced in the other files, new errors can creep in the ignored files
- If/when the errors in the ignored files get removed, the user has to remember to re-enable the rule on this file. Otherwise new errors can creep in also.

### What do you think is the correct solution?

Here is some prior art on suppressing existing errors:
- [`elm-review`](https://elm-review.com)'s suppression system - described in this article https://jfmengels.net/stop-the-bleed/
- [Betterer](https://phenomnomnominal.github.io/betterer/)
- [`ember-template-lint`'s TODO system](https://github.com/ember-template-lint/ember-template-lint/blob/master/docs/todos.md)

I am the author of the linter and blog post mentioned in the first bullet point. My recommendation would be to create a similar system to this one. It has worked really well so far for my users and I.
The blog post also describes the problems I've mentioned in greater detail than in this issue.

A similar discussion was opened in the Ruff linter: https://github.com/astral-sh/ruff/issues/1149

I am not willing to submit a PR nor a RFC as I am not using Biome and this would be a large endeavor, but I'd be happy to give advice and recommendations on how such a system should be put in place.

(PS: If you like this kind of proposal, I have plenty of other ideas on how to make Biome a better linter: https://www.youtube.com/watch?v=XjwJeHRa53A, happy to discuss them elsewhere)

**Comment by Netail** [comment]
Something like https://eslint.org/blog/2025/04/introducing-bulk-suppressions/

**Reply by jfmengels** [reply]
Yes. I made the exact same proposal there (see the first line of my post), which turned into that feature :slightly_smiling_face:

**Reply by Netail** [reply]
Love it

**Comment by eMerzh** [comment]
FTR oxc seems to move supporting eslint's supressions... https://github.com/oxc-project/oxc/pull/19328 might be nice to have something similar and compatible

**Comment by dyc3** [comment]
FYI, this exists now in the form of the `--suppress` argument, which allows you to automatically add suppression comments to all lint errors. 

Simply run this command: 
```bash
biome lint --suppress --reason "initial biome migration"
```

The way our suppression comments work is that you are required to provide a reason, otherwise we emit a warning. So you can easily distinguish between suppressions that were a part of your switch to biome, and suppressions that have other more nuanced reasons.

**Reply by eMerzh** [reply]
although the suppress is a nice inbetween solution, it's a bit different IMHO... you stil have to make sure all your comments are the sames between supression, you'll have to grep /... to have a list of suppression to fix

**Reply by eMerzh** [reply]
also, if a rule change name (nursery or plugins for example) then you need to change all your comments
